plugin/builtin/prefixtransformer/PrefixTransformer.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+//go:generate pluginator
+package main
+
+import (
+	"errors"
+	"sigs.k8s.io/kustomize/api/filters/prefix"
+
+	"sigs.k8s.io/kustomize/api/resmap"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/resid"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Add the given prefix to the field
+type plugin struct {
+	Prefix     string        `json:"prefix,omitempty" yaml:"prefix,omitempty"`
+	FieldSpecs types.FsSlice `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+}
+
+//noinspection GoUnusedGlobalVariable
+var KustomizePlugin plugin
+
+var prefixFieldSpecsToSkip = types.FsSlice{
+	{Gvk: resid.Gvk{Kind: "CustomResourceDefinition"}},
+	{Gvk: resid.Gvk{Group: "apiregistration.k8s.io", Kind: "APIService"}},
+	{Gvk: resid.Gvk{Kind: "Namespace"}},
+}
+
+func (p *plugin) Config(
+	_ *resmap.PluginHelpers, c []byte) (err error) {
+	p.Prefix = ""
+	p.FieldSpecs = nil
+	err = yaml.Unmarshal(c, p)
+	if err != nil {
+		return
+	}
+	if p.FieldSpecs == nil {
+		return errors.New("fieldSpecs is not expected to be nil")
+	}
+	return
+}
+
+func (p *plugin) Transform(m resmap.ResMap) error {
+	// Even if the Prefix is empty we want to proceed with the
+	// transformation. This allows to add contextual information
+	// to the resources (AddNamePrefix).
+	for _, r := range m.Resources() {
+		// TODO: move this test into the filter (i.e. make a better filter)
+		if p.shouldSkip(r.OrgId()) {
+			continue
+		}
+		id := r.OrgId()
+		// current default configuration contains
+		// only one entry: "metadata/name" with no GVK
+		for _, fs := range p.FieldSpecs {
+			// TODO: this is redundant to filter (but needed for now)
+			if !id.IsSelected(&fs.Gvk) {
+				continue
+			}
+			// TODO: move this test into the filter.
+			if fs.Path == "metadata/name" {
+				// "metadata/name" is the only field.
+				// this will add a prefix to the resource
+				// even if it is empty
+
+				r.AddNamePrefix(p.Prefix)
+				if p.Prefix != "" {
+					r.StorePreviousId()
```

Comment:
This may have unintended consequences, I'm trying to think of the best way to approach this. 

For example, if we have the following in a kustomization:

```
namePrefix: foo-
nameSuffix: -bar
```

and this gets applied to a resource with the name:

```
metadata:
  name: resource
```

we will get the following:
```
metadata:
  name: foo-resource-bar
  annotations:
    internal.config.kubernetes.io/previousNames: resource, foo-resource
```

Because we are separating the prefix transformer and the suffix transformer, we are now making two calls to `StorePreviousID()`. This is a very subtle change, but it is possible for it to cause problems. However, we already have this problem... the `patchTransformer` also can change a resource's name and makes a call to `StorePreviousID()`, so since you are not introducing a problem we don't already have, I think it will be okay to accept the change.

Would you mind adding a comment both here and in SuffixTransformer.go as follows:
```
// TODO: There are multiple transformers that can change a resource's name, and each makes a call to 
// StorePreviousID(). We should make it so that we only call StorePreviousID once per kustomization layer 
// to avoid storing intermediate names between transformations, to prevent intermediate name conflicts. 
```

Reply:
I was confused about what `StorePreviousID()` exactly does but now I understand. I will add the TODOs 👍🏼